support-lib/cpp/Future.hpp

```diff
@@ -0,0 +1,307 @@
+        assert(sharedState);    // call on invalid future will trigger assertion
+        std::unique_lock lk(sharedState->mutex);
+        sharedState->cv.wait(lk, [state = sharedState] {return state->isReady();});
+        if (!sharedState->exception) {
+            return sharedState->getValueUnsafe();
+        } else {
+            std::rethrow_exception(sharedState->exception);
+        }
+    }
+    // If at the moment of calling `then()`, the result (or exception) is
+    // already available, then the handler routine will immediately be called in
+    // the current thread. Returns a new future that wraps the return value of
+    // the handler routine. The current future becomes invalid after this call.
+    template<typename FUNC>
+    auto then(FUNC&& handler) {
+        using HandlerReturnType = std::invoke_result_t<FUNC, Future<T>>;
+        detail::SharedStatePtr<T> sharedState;
+        sharedState = std::atomic_exchange(&_sharedState, sharedState);
+        assert(sharedState);    // a second call will trigger assertion
+        auto nextPromise = std::make_unique<Promise<HandlerReturnType>>();
+        auto nextFuture = nextPromise->getFuture();
+        auto continuation = [handler = std::forward<FUNC>(handler), nextPromise = std::move(nextPromise)] (detail::SharedStatePtr<T> x) mutable {
+            try {
+                if constexpr(std::is_void_v<HandlerReturnType>) {
+                    handler(Future<T>(x));
+                    nextPromise->setValue();
+                } else {
+                    nextPromise->setValue(handler(Future<T>(x)));
+                }
+            } catch (std::exception& e) {
+                nextPromise->setException(std::current_exception());
+            }
+        };
+        detail::SharedStatePtr<T> sharedStateForReadyFuture;
+        {
+            std::unique_lock lk(sharedState->mutex);
+            if (sharedState->isReady()) {
+                // result already available
+                sharedStateForReadyFuture = std::move(sharedState);
+            } else {
+                // result not yet available
+                sharedState->handler = detail::createValueHandler<T>(std::move(continuation));
+            }
+        }
+        if (sharedStateForReadyFuture) {
+            continuation(sharedStateForReadyFuture);
+        }
+        return nextFuture;
+    }
+
+private:
+    detail::SharedStatePtr<T> _sharedState;
+
+#if defined(DJINNI_FUTURE_COROUTINE_SUPPORT)
+public:
+    bool await_ready() {
+        return isReady();
+    }
+    auto await_resume() {
+        auto sharedState = std::atomic_load(&_sharedState);
+        if (sharedState) {
+            return Future<T>(sharedState).get();
+        } else {
+            return Future<T>(_coroState).get();
+        }
+    }
+    bool await_suspend(std::experimental::coroutine_handle<> h) {
+        this->then([h, this] (Future<T> x) mutable {
+            _coroState = x._sharedState;
+            h();
```

Comment:
H.destroy();

Reply:
actually, no.  our coroutine support library returns `std::suspend_never` in its `final_suspend()` method. so the handle is automatically destroyed when the coroutine returns.